BUG: preserve name in set_categories (#17509)

Giftlin opened the pull request:
- [x] closes #17509
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Giftlin committed:
FIX for set_categories issue #17509

TomAugspurger commented on pandas/core/categorical.py:
```diff
@@ -2068,7 +2068,7 @@ def _delegate_method(self, name, *args, **kwargs):
         res = method(*args, **kwargs)
         if res is not None:
-            return Series(res, index=self.index)
+            return Series(res, index=self.index, name=self.name)
```
Ah, I misread the earlier. `self` is the instance of `CategoricalAccessor`, not the `Series` like I expected. So this approach won't work.

Giftlin replied:
so, what do you think will work for this scenario?

TomAugspurger replied:
I think if you modify https://github.com/pandas-dev/pandas/blob/f11bbf2f505d81900cc83ce387a6a1b1d2a2f866/pandas/core/categorical.py#L2078 to get `data.name`. and then modify `CategoricalAccessor.__init__` to accept a name. `data` may not have a name, so maybe `getattr(data, 'name', None)`.

jreback replied:
so the signature should follow like we do in ``pandas.core.indexes.accessors``, e.g. ``(data, index, name=None)``

TomAugspurger commented:
Also, if you could add new tests to ensure the name is passed through, and a release note in doc/source/whatsnew/v0.21.0.txt under bug fixes.

Giftlin committed:
FIX for set_categories issue #17509

Giftlin committed:
FIX for set_categories issue #17509

pep8speaks commented:
Hello @Giftlin! Thanks for updating the PR.

 - In the file [`pandas/core/categorical.py`](https://github.com/pandas-dev/pandas/blob/4f731cffcdd6ad0e302ac09603cd21526e627c88/pandas/core/categorical.py), following are the PEP8 issues :

> [Line 2079:80](https://github.com/pandas-dev/pandas/blob/4f731cffcdd6ad0e302ac09603cd21526e627c88/pandas/core/categorical.py#L2079): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (88 > 79 characters)

Giftlin committed:
Update categorical.py

jreback reviewed (changes requested):
tests!

Giftlin commented:
@jreback I'm really sorry, this is my first time trying to fix, if u don't mind, can u give me a sample and guide me? :(

Giftlin committed:
Update categorical.py

Giftlin commented:
#17529

jreback commented:
contrib docs are [here](http://pandas.pydata.org/pandas-docs/stable/contributing.html). pls add tests IN this PR. It should fail with the tests, then pass after the fix.

Giftlin committed:
Test for the changes

Giftlin committed:
Merge pull request #2 from Giftlin/patch-7

Patch 7

Giftlin committed:
Space removed

Giftlin commented:
@jreback , tests done.. 👍

Giftlin committed:
- Bug in preserving name in set_categories. (:issue:`17509`)

Giftlin committed:
Merge pull request #3 from Giftlin/patch-10

- Bug in preserving name in set_categories. (:issue:`17509`)

Giftlin commented:
@jorisvandenbossche 
@jreback 
@TomAugspurger 
Tests done.. can someone please review and merge this? :(

TomAugspurger commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -566,6 +566,7 @@ Categorical
 - Bug in the categorical constructor with empty values and categories causing
   the ``.categories`` to be an empty ``Float64Index`` rather than an empty
   ``Index`` with object dtype (:issue:`17248`)
+- Bug in preserving name in ``set_categories``. (:issue:`17509`)
```
Could you make this more descriptive? Maybe

    Bug in categorical operations on Series like `Series.cat.set_categories` not preserving the original Series' name (:issue:`17509`)

TomAugspurger commented on pandas/tests/test_categorical.py:
```diff
@@ -57,6 +57,13 @@ def test_getitem_listlike(self):
         tm.assert_numpy_array_equal(result, expected)
 
+    def test_getname_category(self):
+        result = 'A'
```
`result` should be the result of the operation. So. `result = s.cat.set_categories([1, 2, 3]).name`. And then compare directly to the expected name. `assert result == 'A'`

TomAugspurger commented on pandas/tests/test_categorical.py:
```diff
@@ -57,6 +57,13 @@ def test_getitem_listlike(self):
+        s = s.cat.set_categories([1, 2, 3])
+        expected = s.astype('category').name
+        tm.assert_almost_equal(result, expected)
+
```
Can you add additional tests for the other delegated methods that this should apply to? `set_ordered`, etc?

Giftlin committed:
Update v0.21.0.txt

Giftlin committed:
Update test_categorical.py

TomAugspurger commented on pandas/tests/test_categorical.py:
```diff
@@ -57,6 +57,13 @@ def test_getitem_listlike(self):
+        expected = 'A'
+        s = pd.Series([1, 2, 3], name='A').astype('category')
+        s = s.cat.set_categories([1, 2, 3])
+        result = s.astype('category').name
```
This should be `result == s.cat.set_categories([1, 2, 3]).name` so that we're directly testing the behavior that caused the bug.

TomAugspurger commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -566,7 +566,9 @@ Categorical
   the ``.categories`` to be an empty ``Float64Index`` rather than an empty
   ``Index`` with object dtype (:issue:`17248`)
-
+- Bug in categorical operations on Series like `Series.cat.set_categories`
```
Sorry, it should be double backticks, not single

    ``Series.cat.set_categories``

Giftlin committed:
Update v0.21.0.txt

Giftlin committed:
Update test_categorical.py

jreback commented on pandas/tests/test_categorical.py:
```diff
@@ -57,6 +57,12 @@ def test_getitem_listlike(self):
         expected = c[np.array([100000]).astype(np.int64)].codes
         tm.assert_numpy_array_equal(result, expected)
 
+    def test_getname_category(self):
```
parametrize with all accessor methods

jreback replied:
you can use a lambda expression

```
@pytest.mark.parametrize("method",
[
   lambda x: x.cat.set_categories([1, 2,3 ]),
   lambda x: x.cat.reorder_categories([2,3, 1], ordered=True),
])
def test_getname_categorical_accessor(self, method):
     s = ....
    expected = 'A'
    result = method(s).name
    assert result == expected
```

Giftlin replied:
The current test function is also fine only right? What is the difference? Performance?

Giftlin committed:
Update test_categorical.py

jreback commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -566,7 +566,9 @@ Categorical
   the ``.categories`` to be an empty ``Float64Index`` rather than an empty
   ``Index`` with object dtype (:issue:`17248`)
-
+- Bug in categorical operations on Series like ``Series.cat.set_categories``
```
should instead refer to all ``Series.cat`` operations not preserving name

Giftlin replied:
Should I mention all the functions along with set_categories here?

jreback replied:
I would just refer to http://pandas.pydata.org/pandas-docs/stable/categorical.html#working-with-categories

Giftlin replied:
@jreback , the tests have passed but Travis CI build has failed. Y is that? :(

Giftlin committed:
Update v0.21.0.txt

Giftlin committed:
Update v0.21.0.txt

jreback commented:
@Giftlin FYI you don't need to push after every commit. you can still make seperate commits, but uless its an isolated change, this just runs the CI more.

Giftlin commented:
@jreback thanks for d info. :) 
I'm very new here. 13 days old . Haven't gotten used to yet.

jreback commented:
> I'm very new here. 13 days old . Haven't gotten used to yet.

np. thanks for the contributions! you are doing great!

Giftlin committed:
Update test_categorical.py

Giftlin commented:
@jreback , tests passed. Is the test function fine? Or I should parameterize it?

jreback commented:
parameteize

jreback commented:
and u have some lint issues

Giftlin committed:
Merge pull request #4 from Giftlin/Giftlin-param_test_categorical.py

parametrizing with all accessor methods

Giftlin commented:
How do I fix the issues?

Giftlin committed:
Update v0.21.0.txt

Giftlin commented:
I'm not sure what causes the lint issues. Can u help me out here? :(

Giftlin committed:
Lint error fix

Giftlin committed:
Merge pull request #5 from Giftlin/Giftlin-lint-error-patch

Lint error fix

Giftlin committed:
Removed space

Sorry for too many commits.

Giftlin committed:
Lint error

Giftlin committed:
Merge pull request #6 from Giftlin/Giftlinlint

Lint error

jreback commented on pandas/tests/test_categorical.py:
```diff
@@ -57,6 +57,24 @@ def test_getitem_listlike(self):
         expected = c[np.array([100000]).astype(np.int64)].codes
         tm.assert_numpy_array_equal(result, expected)
 
+    @pytest.mark.parametrize("method",
```
not sure if this is passing linting I would write this like

```
@pytest.mark.parametrize(
    "method",
    [
        lambda .....
    ])
```

to get more indent

jreback commented:
ok this looks fine to me (after fixing linting). ping on green.

Giftlin committed:
Update test_categorical.py

Giftlin committed:
Update test_categorical.py

Giftlin committed:
Update test_categorical.py

Giftlin commented:
@jreback, fixed limiting.. checks have passed

jorisvandenbossche commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -566,6 +566,7 @@ Categorical
 - Bug in the categorical constructor with empty values and categories causing
   the ``.categories`` to be an empty ``Float64Index`` rather than an empty
   ``Index`` with object dtype (:issue:`17248`)
+- Bug in categorical operations `Series.cat <http://pandas.pydata.org/pandas-docs/stable/categorical.html#working-with-categories>' not preserving the original Series' name (:issue:`17509`)
```
Sphinx has the ability to do local references (see http://www.sphinx-doc.org/en/stable/markup/inline.html#cross-referencing-arbitrary-locations), so you can do `` :ref:`Series.cat <categorical.cat>`   `` instead of listing the actual full url. 
You only need to add a label at the specific location (put `` .. _categorical.cat: `` on the line before the title):

https://github.com/pandas-dev/pandas/blob/cbb090fac6a39d0e687a01b2ad1b7c136f5d92fc/doc/source/categorical.rst#L148-L150

Giftlin replied:
This is the expected change right?

![screenshot_2017-09-18-13-25-56-574_com android browser](https://user-images.githubusercontent.com/31629119/30533041-fbc6b006-9c74-11e7-89e8-ccb474aeb25c.png)

jorisvandenbossche replied:
yes!

jorisvandenbossche reviewed (commented):
Looks good! 
Just a minor comment about how to link to a section in the docs.

Giftlin committed:
Update v0.21.0.txt

http://pandas.pydata.org/pandas-docs/stable/categorical.html#working-with-categories

Giftlin committed:
Update v0.21.0.txt

Giftlin committed:
Update categorical.rst

Giftlin committed:
Merge pull request #9 from Giftlin/Giftlin-patch-1-1

Update categorical.rst

Giftlin committed:
Merge pull request #7 from Giftlin/Giftlin-whatsnew

What's new

gfyoung commented on pandas/tests/test_categorical.py:
```diff
@@ -57,6 +57,24 @@ def test_getitem_listlike(self):
+            lambda x: x.cat.as_ordered(),
+            lambda x: x.cat.as_unordered(),
+        ])
+    def test_getname_categorical_accessor(self, method):
```
Reference issue below function definition.

gfyoung replied:
Changed acknowledged.  Thanks!

gfyoung commented on pandas/tests/test_categorical.py:
```diff
@@ -57,6 +57,24 @@ def test_getitem_listlike(self):
+        s = pd.Series([1, 2, 3], name='A').astype('category')
+        expected = 'A'
+        result = method(s).name
+        assert result == expected
```
Instead of doing this, construct the expected `Series` directly from the constructor and call `tm.assert_series_equal`.

Giftlin replied:
But Series is not expected. Only the name is expected, which is a character

gfyoung replied:
That's because you defined your test as such.  The output of `method` is a `Series`, not a character.  I'm saying that you should check the output of `method` instead of comparing its `name` attribute.

jorisvandenbossche replied:
@gfyoung In this case asserting only the name is fine I think. Assuming the actual methods are already tested separately, this tests just asserts for all of them they preserve the name. If checking the actual resulting series, all of them would need a different result, so would just over-complicate the test.

Giftlin committed:
Update test_categorical.py

Giftlin committed:
Merge pull request #10 from Giftlin/Giftlin-test_categorical.py-1

Update test_categorical.py

Giftlin commented:
@jreback , all checks have passed

jorisvandenbossche committed:
small edits

Giftlin commented:
@jreback
@TomAugspurger 
checks have passed

jorisvandenbossche commented:
@Giftlin Thanks! 

(I added a small commit with some white-space clean-up. Not sure what went wrong, but the diff showed changes in whitespace on some lines. You might need to check the settings of your editor to ensure it only uses spaces for whitespace)
